### PR TITLE
Add methods to BotAware along with moving docs

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
Add methods around starting and stopping the bot app to the BotAware trait and move docs from impl

### Checklist

- [ x] I have run `nox` and all the pipelines have passed.
- [x ] I have made unittests according to the code I have added/modified/deleted.

### CODE REVIEW

File: hikari/traits.py [hunk 1]
```diff
@@ -446,4 +446,177 @@ async def update_presence(
...
+        close_executor: bool = False,
+        close_loop: bool = True,
+        coroutine_tracking_depth: typing.Optional[int] = None,
+        enable_signal_handlers: bool = True,
+        idle_since: typing.Optional[datetime.datetime] = None,
+        ignore_session_start_limit: bool = False,
+        large_threshold: int = 250,
+        status: presences.Status = presences.Status.ONLINE,
+        shard_ids: typing.Optional[typing.Set[int]] = None,
+        shard_count: typing.Optional[int] = None,
```
Comment (nekokatt) [inline hikari/traits.py:483]: I don't agree that most of these should be present.

Half of this stuff is implementation specific, and implementations may not want to or may not be able to provide some of these.

I'd revert:
- asyncio_debug  (as their loop may not support this properly or may have other implications if it uses an existing loop)
- close_loop  (as their impl may use the event loop outside this)
- coroutine_tracking_depth (provisional feature in CPython, may go at any time)
- enable_signal_handlers (implementations may not bother with signals at all)

**Review by nekokatt — CHANGES_REQUESTED** [review]

File: hikari/traits.py (hunk 1 above)
Comment (FasterSpeeding) [reply hikari/traits.py:483]: ok 👍

Comment (nekokatt) [reply hikari/traits.py:483]: Under the rule of `def x(a, b, c)` being compatible with `def x(a, b, c, d=None, e=None, f=None), this should be fine with MyPy.

Comment (nekokatt) [reply hikari/traits.py:483]: Also afaik unfortunately you'll have to duplicate this docstring because pdoc3 defaults :/

Comment (FasterSpeeding) [reply hikari/traits.py:483]: Well if we're changing the args from protocol to impl it won't be duping anyway and pdoc3 is inheriting the docs here

Comment (FasterSpeeding) [reply hikari/traits.py]: Removed the ones you listed, if you think any more should be removed then send a list

Comment (nekokatt) [reply hikari/traits.py]: thats it

**Review by nekokatt — APPROVED** [review]

**Comment by nekokatt** [comment]
Please can you rebase onto master.

Chat to @davfsa if unsure about changes

File: hikari/traits.py [hunk 2]
```diff
@@ -446,4 +446,146 @@ async def update_presence(
...
+            Defaults to `builtins.False`. If `builtins.True`, any custom
+            `concurrent.futures.Executor` passed to the constructor will be
+            shut down when the application terminates. This does not affect the
+            default executor associated with the event loop, and will not
+            do anything if you do not provide a custom executor to the
+            constructor.
+        """
+        raise NotImplementedError
+
     __slots__: typing.Sequence[str] = ()
```
Comment (nekokatt) [inline hikari/traits.py:591]:
```suggestion
```

File: hikari/traits.py [hunk 3]
```diff
@@ -446,4 +446,146 @@ async def update_presence(
 class BotAware(RESTAware, ShardAware, EventFactoryAware, DispatcherAware, typing.Protocol):
     """Structural supertype for a component that is aware of all internals."""
 
+    async def join(self, until_close: bool = True) -> None:
```
Comment (nekokatt) [inline hikari/traits.py:449]:
```suggestion
class BotAware(RESTAware, ShardAware, EventFactoryAware, DispatcherAware, typing.Protocol):
    """Structural supertype for a component that is aware of all internals."""
    
    __slots__: typing.Sequence[str] = ()

    async def join(self, until_close: bool = True) -> None:
```

**Review by nekokatt — CHANGES_REQUESTED** [review]
Fixing conflicts with master and fixing a slots that fell out of the right place :)

File: hikari/traits.py [hunk 4]
```diff
@@ -446,4 +446,129 @@ async def update_presence(
 class BotAware(RESTAware, ShardAware, EventFactoryAware, DispatcherAware, typing.Protocol):
     """Structural supertype for a component that is aware of all internals."""
 
-    __slots__: typing.Sequence[str] = ()
+    async def join(self, until_close: bool = True) -> None:
+        """Wait indefinitely until the application closes.
```
Comment (nekokatt) [inline hikari/traits.py:450]:
```suggestion
class BotAware(RESTAware, ShardAware, EventFactoryAware, DispatcherAware, typing.Protocol):
    """Structural supertype for a component that is aware of all internals."""

    __slots__: typing.Sequence[str] = ()
    
    async def join(self, until_close: bool = True) -> None:
        """Wait indefinitely until the application closes.
```

Comment (nekokatt) [reply hikari/traits.py:450]: Gah, GitHub won't let me do it. Would you mind just re-adding the slots in for me here? You'd moved them down below everything else and I was trying to undo this for you so you didn't have to waste time fixing stuff that wasn't your fault, but I guess I made it worse. Sorry...

**Review by nekokatt — APPROVED** [review]

**Pull request merged**